tron/kubernetes.py

```diff
@@ -91,3 +104,206 @@ def handle_event(self, event: Event) -> None:
+
+        # creating a task_proc executor has a couple steps:
+        # * create a TaskProcessor
+        # * load the desired plugin (in this case, the k8s one)
+        # * and then actually create the executor (which we call a runner in tron)
+        # this last step requires a bit of setup, which is why we don't do it in-line
+        # in this constructor
+        self.processor = TaskProcessor()
+        self.processor.load_plugin(provider_module="task_processing.plugins.kubernetes")
+        self.runner: Optional[Subscription] = None
+
+        # queue to to use for tron<->task_proc communication - will hold k8s events seen
+        # by task_processing and held for tron to process.
+        self.queue = PyDeferredQueue()
+        # this will hold the current event to process (retrieved from the PyDeferredQueue above)
+        # which we will eventually wrap with some callbacks to actually process using the Twisted
+        # reactor started as part of tron's startup process
+        self.deferred: Optional[Deferred] = None
+
+        # map from k8s pod names to the task that said pod corresponds to
+        self.tasks: Dict[str, KubernetesTask] = {}
+
+        # actually create the executor/runner, as mentioned above.
+        self.connect()
+
+    def connect(self) -> None:
+        """
+        Starts running our Kubernetes task_processing.
+        """
+        self.runner = self.get_runner(kubeconfig_path=self.kubeconfig_path, queue=self.queue)
+        self.handle_next_event()
+
+    def get_runner(self, kubeconfig_path: str, queue: PyDeferredQueue) -> Optional[Subscription]:
+        """
+        Gets or creates an instance of our Kubernetes task_processing plugin.
+        """
+        if not self.enabled:
+            log.info("Kubernetes usage is disabled, not creating a runner.")
+            return None
+
+        # TODO: Add a stopping/terminating state to the task_proc runner
+        if self.runner is not None:
+            log.info("Reusing previously created runner.")
+            return self.runner
+
+        # TODO: once we start implementing more things in the executor, we'll need to actually pass
+        # down some config
+        executor = self.processor.executor_from_config(provider="kubernetes", provider_config={})
+
+        # TODO: we should return a Subscription here: but this will crash until we've actually implemented
+        # get_event_queue() in our task_proc plugin
+        return executor
+
+    def handle_next_event(self, _=None) -> None:
+        """
+        Pops events off of the shared tron<->task_proc queue and processes them.
+
+        We only expect two types of events:
+        * control: events regarding how the task_proc plugin is running - handled directly
+        * task: events regarding how the actual tasks/Pods we're running our doing - forwarded to KubernetesTask
+        """
+        if self.deferred and not self.deferred.called:
```

Comment:
Are we checking if self.deferred has been called? I'd assume it to be true if we already have handlers waiting. When does self.deferred.called become true?

Reply:
self.deferred will either be `None` or a `PyDeferred` object (thus the `if self.deferred`), and until one of the {Call,Err}backs has been called, `self.deferred.call` will be `False` - in this case, `called` will be `False` until `process_event()` or `logError` have been called